src/backend/x11/mod.rs

```diff
@@ -0,0 +1,563 @@
+use x11rb::x11_utils::X11Error as ImplError;
+use x11rb::{atom_manager, protocol as x11};
+
+/// An error that may occur when initializing the backend.
+#[derive(Debug, thiserror::Error)]
+pub enum InitializationError {}
+
+/// An error emitted by the X11 backend.
+#[derive(Debug, thiserror::Error)]
+pub enum X11Error {
+    /// An error that may occur when initializing the backend.
+    #[error("Error while initializing backend")]
+    Connect(ConnectToXError),
+
+    /// Connecting to the X server failed.
+    #[error("Connecting to the X server failed")]
+    ConnectionFailed(ConnectError),
+
+    /// An X11 error packet was encountered.
+    #[error("An X11 error packet was encountered.")]
+    Protocol(ReplyOrIdError),
+
+    /// The window was destroyed.
+    #[error("The window was destroyed")]
+    WindowDestroyed,
+}
+
+impl From<ConnectToXError> for X11Error {
+    fn from(e: ConnectToXError) -> Self {
+        X11Error::Connect(e)
+    }
+}
+
+impl From<ConnectError> for X11Error {
+    fn from(e: ConnectError) -> Self {
+        X11Error::ConnectionFailed(e)
+    }
+}
+
+impl From<ConnectionError> for X11Error {
+    fn from(e: ConnectionError) -> Self {
+        ReplyOrIdError::from(e).into()
+    }
+}
+
+impl From<ImplError> for X11Error {
+    fn from(e: ImplError) -> Self {
+        ReplyOrIdError::from(e).into()
+    }
+}
+
+impl From<ReplyError> for X11Error {
+    fn from(e: ReplyError) -> Self {
+        ReplyOrIdError::from(e).into()
+    }
+}
+
+impl From<ReplyOrIdError> for X11Error {
+    fn from(e: ReplyOrIdError) -> Self {
+        X11Error::Protocol(e)
+    }
+}
+
+/// Properties defining information about the window created by the X11 backend.
+#[derive(Debug, Clone, Copy)]
+#[allow(missing_docs)] // Self explanatory fields
+pub struct WindowProperties<'a> {
+    pub width: u16,
+    pub height: u16,
+    pub title: &'a str,
+}
+
+impl Default for WindowProperties<'_> {
+    fn default() -> Self {
+        WindowProperties {
+            width: 1280,
+            height: 800,
+            title: "Smithay",
+        }
+    }
+}
+
+/// An event emitted by the X11 backend.
+#[derive(Debug)]
+pub enum X11Event {
+    /// The X server has required the compositor to redraw the window.
+    Refresh,
+
+    /// An input event occurred.
+    Input(InputEvent<X11Input>),
+
+    /// The window was resized.
+    Resized(Size<u16, Logical>),
+
+    /// The window was requested to be closed.
+    CloseRequested,
+}
+
+// TODO:
+// Figure out how to expose this to the outside world after inserting in the event loop so buffers can be allocated?
+
+/// An abstraction representing a connection to the X11 server.
+#[derive(Debug)]
+pub struct X11Backend {
+    log: Logger,
+    connection: Arc<XConnection>,
+    source: X11Source,
+    window: Arc<WindowInner>,
+    key_counter: Arc<AtomicU32>,
+    depth: Depth,
+    visual_id: u32,
+    gbm_device: Device<RawFd>,
+}
+
+unsafe impl Send for X11Backend {}
```

Comment:
Uhm... this might need a comment. Is this about `gbm_device`? Why is it not `Send` and why is your usage fine?

Why do you need to send the X11Backend between threads anyway?

Reply:
I think this was a relic of the tiny attempt I made with the egl context, but I realized it's not the best approach. Probably should remove it